orm/versioning_bucket.go

```diff
@@ -233,3 +226,40 @@ func (marker) Copy() CloneableData {
 func (m marker) Equal(o []byte) bool {
 	return bytes.Equal(m, o)
 }
+
+// MarshalVersionedID is used to guarantee determinism while serializing a VersionedIDRef.
+// It comes with the option to omit empty version should you want to do a prefix query.
+// This flag should be set to false when writing to the underlying storage as we depend
+// on the fact that last 4 bytes of the IDRef are used for version in UnmarshalVersionedID.
+func MarshalVersionedID(key VersionedIDRef, omitEmptyVersion bool) []byte {
+	res := make([]byte, 0, len(key.ID)+versionMaxBytes)
+
+	res = append(res, key.ID...)
+
+	if omitEmptyVersion && key.Version == 0 {
```

Comment:
This API can be simplified. ~Versioning starts with 1. This means that `omitEmptyVersion` is not needed. You can use `version == 0` to return key without version. In all cases where `version > 0` returned key should contain the version information.~
If you need only the ID, you can use `VersionedIDRef.ID`. Otherwise you would use `MarshalVersionedID` to serialize the whole reference, together with the version (regardless the value). `omitEmptyVersion` is not needed, because both functionalities can be provided without it.

In the current version `MarshalVersionedID` is not compatible with `UnmarshalVersionedID`.

Reply:
This was done on purpose, if you look at the commit history I did start with using just the ID, but the I have figured out that for all the external dependencies it is nicer to use this method than doing
```
if id.Version == 0 { .... }
```

It is easier to have a flag in this method than having several places in the code that are using `ID` directly for prefix searches.

While I'm not a big fan of this change I think it is cleaner and more straightforward to use for anyone who might want to utilise these two methods, as both of them are well documented. 

The alternative was to introduce a 3rd method, but that's one too many.

Reply:
I am with Piotr on this. `MarshalVersionedID/UnmarshalVersionedID` outputs should be usable to feed each other. This additional `omitEmptyVersion` does not make it more readable to me.
It is only set `true` in the bnscli client and a `gov/bucket_test`
The test could use any random []byte slice to achieve the same results as it is "By unknown key". Just move the related code to bnscli and all good.